I'm not sure this is going to work.
```py
>>> # reproducing the original error
>>> e = http.Http404('\xef\xbd\xa1')
>>> unicode(e)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'ascii' codec can't decode byte 0xef in position 0: ordinal not in range(128)

>>> # the fix doesn't seem to work
>>> unicode(e, encoding='utf-8')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: coercing to Unicode: need string or buffer, Http404 found

>>> # this works
>>> str(e).decode('utf-8')
u'\uff61'

>>> # but it doesn't work for a unicode exception message
>>> f = http.Http404('\xef\xbd\xa1'.decode('utf-8'))
>>> str(f).decode('utf-8')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeEncodeError: 'ascii' codec can't encode character u'\uff61' in position 0: ordinal not in range(128)
```
How hard would it be to write tests for this?

--- a/corehq/util/tests/__init__.py
+++ b/corehq/util/tests/__init__.py
@@ -4,6 +4,7 @@ from test_couchdb_management import *
 from test_datadog_utils import *
 from test_global_request import *
 from test_jsonobject import *
+from test_json_error import *
 from test_log import *
 from test_override_db import *
 from test_quickcache import *

--- a/corehq/util/tests/test_json_error.py
+++ b/corehq/util/tests/test_json_error.py
@@ -1,0 +1,18 @@
+from django.test.testcases import SimpleTestCase
+
+from corehq.util.test_utils import generate_cases
+from corehq.util.view_utils import _json_exception_response_data
+
+
+class JsonErrorTests(SimpleTestCase):
+    """Test json_error decorator"""
+
+
+@generate_cases([
+    ('ascii string',),
+    ('utf8 string \xef\xbd\xa1', u'utf8 string \uff61'),
+    (u'unicode string \uff61',),
+], JsonErrorTests)
+def test_json_exception_response_data(self, message, expected=None):
+    data = _json_exception_response_data(500, Exception(message))
+    self.assertEqual(data['message'], expected or message)

--- a/corehq/util/view_utils.py
+++ b/corehq/util/view_utils.py
@@ -62,47 +62,46 @@ def json_error(f):
             return response
 
         except http.Http404 as e:
-            blob = json.dumps({
-                'error': 404,
-                'message': unicode(e),
-            })
-            logger.warning('Not found: %s', request.path,
-                           extra={
-                               'status_code': 404,
-                               'request': request,
-                           })
-            return http.HttpResponseNotFound(blob, content_type=JSON)
+            return _get_json_exception_response(404, request, e, log_message='Not found: %s')
         except PermissionDenied as e:
-            logger.warning(
-                'Forbidden (Permission denied): %s', request.path,
-                extra={
-                    'status_code': 403,
-                    'request': request,
-                })
-            blob = json.dumps({
-                'error': 403,
-                'message': unicode(e),
-            })
-            return http.HttpResponseForbidden(blob, content_type=JSON)
+            return _get_json_exception_response(403, request, e, log_message='Forbidden (Permission denied): %s')
         except BadRequest as e:
-            blob = json.dumps({
-                'error': 400,
-                'message': unicode(e),
-            })
-            return http.HttpResponseBadRequest(blob, content_type=JSON)
+            return _get_json_exception_response(400, request, e)
         except Exception as e:
-            data = {
-                'error': 500,
-                'message': unicode(e)
-            }
-            if settings.DEBUG:
-                data['traceback'] = traceback.format_exc()
-            return http.HttpResponse(
-                status=500,
-                content=json.dumps(data),
-                content_type=JSON
-            )
+            return _get_json_exception_response(500, request, e)
     return inner
+
+
+def _get_json_exception_response(code, request, exception, log_message=None):
+    if log_message:
+        logger.warning(
+            log_message, request.path,
+            extra={
+                'status_code': code,
+                'request': request,
+            })
+
+    data = _json_exception_response_data(code, exception)
+
+    return http.HttpResponse(
+        status=code,
+        content=json.dumps(data),
+        content_type=JSON
+    )
+
+
+def _json_exception_response_data(code, exception):
+    if isinstance(exception.message, unicode):
+        message = unicode(exception)
+    else:
+        message = str(exception).decode('utf-8')
+    data = {
+        'error': code,
+        'message': message
+    }
+    if code == 500 and settings.DEBUG:
+        data['traceback'] = traceback.format_exc()
+    return data
 
 
 def get_request():

